File: src/go/airlock/crypto.go
package airlock

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdh"
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"fmt"
	"io"

	"golang.org/x/crypto/hkdf"
)

const (
	aesKeySize   = 32
	aesNonceSize = 12
	aesTagSize   = 16
)

// EncryptedPayload represents an AES-256-GCM encrypted payload.
// Wire-compatible with the CiphertextRef schema in the HARP Gateway spec.
type EncryptedPayload struct {
	Alg   string `json:"alg"`
	Data  string `json:"data"`
	Nonce string `json:"nonce,omitempty"`
	Tag   string `json:"tag,omitempty"`
	Aad   string `json:"aad,omitempty"`
}

// ── Base64url ────────────────────────────────────────────────

// ToBase64URL encodes bytes to a base64url string (no padding).
func ToBase64URL(data []byte) string {
	return base64.RawURLEncoding.EncodeToString(data)
}

// FromBase64URL decodes a base64url string (with or without padding).
func FromBase64URL(s string) ([]byte, error) {
	// Try without padding first, then with
	data, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		data, err = base64.URLEncoding.DecodeString(s)
	}
	return data, err
}

// ── SHA-256 ──────────────────────────────────────────────────

// SHA256Hex computes the lowercase hex SHA-256 of a string.
func SHA256Hex(s string) string {
	h := sha256.Sum256([]byte(s))
	return hex.EncodeToString(h[:])
}

// ── HKDF-SHA256 ──────────────────────────────────────────────

// HKDFDerive derives a key using HKDF-SHA256 (RFC 5869).
func HKDFDerive(ikm, salt, info []byte, length int) ([]byte, error) {
	r := hkdf.New(sha256.New, ikm, salt, info)
	out := make([]byte, length)
	if _, err := io.ReadFull(r, out); err != nil {
		return nil, fmt.Errorf("hkdf derive: %w", err)
	}
	return out, nil
}

// ── X25519 ECDH Key Exchange ─────────────────────────────────

// X25519KeyPair holds an X25519 keypair for ECDH key agreement.
type X25519KeyPair struct {
	PublicKey  string // base64url (raw 32 bytes)
	PrivateKey string // base64url (raw 32 bytes)
}

// GenerateX25519KeyPair generates an X25519 keypair for ECDH.
func GenerateX25519KeyPair() (*X25519KeyPair, error) {
	curve := ecdh.X25519()
	privKey, err := curve.GenerateKey(rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("generate x25519 key: %w", err)
	}
	return &X25519KeyPair{
		PublicKey:  ToBase64URL(privKey.PublicKey().Bytes()),
		PrivateKey: ToBase64URL(privKey.Bytes()),
	}, nil
}

// DeriveSharedKey derives a shared AES-256 key from X25519 ECDH + HKDF-SHA256.
// Uses info string "HARP-E2E-AES256GCM" to match the enforcer extension pattern.
func DeriveSharedKey(myPrivateKeyBase64URL, peerPublicKeyBase64URL string) (string, error) {
	privBytes, err := FromBase64URL(myPrivateKeyBase64URL)
	if err != nil {
		return "", fmt.Errorf("decode private key: %w", err)
	}
	pubBytes, err := FromBase64URL(peerPublicKeyBase64URL)
	if err != nil {
		return "", fmt.Errorf("decode public key: %w", err)
	}

	curve := ecdh.X25519()
	privKey, err := curve.NewPrivateKey(privBytes)
	if err != nil {
		return "", fmt.Errorf("import private key: %w", err)
	}
	pubKey, err := curve.NewPublicKey(pubBytes)
	if err != nil {
		return "", fmt.Errorf("import public key: %w", err)
	}

	sharedSecret, err := privKey.ECDH(pubKey)
	if err != nil {
		return "", fmt.Errorf("ecdh: %w", err)
	}

	info := []byte("HARP-E2E-AES256GCM")
	derived, err := HKDFDerive(sharedSecret, nil, info, aesKeySize)
	if err != nil {
		return "", err
	}

	return ToBase64URL(derived), nil
}

// ── AES-256-GCM Encrypt/Decrypt ──────────────────────────────

// AesGcmEncrypt encrypts plaintext with AES-256-GCM (detached nonce+tag).
func AesGcmEncrypt(keyBase64URL, plaintext string) (*EncryptedPayload, error) {
	key, err := FromBase64URL(keyBase64URL)
	if err != nil {
		return nil, fmt.Errorf("decode key: %w", err)
	}
	if len(key) != aesKeySize {
		return nil, fmt.Errorf("key must be %d bytes, got %d", aesKeySize, len(key))
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("new cipher: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("new gcm: %w", err)
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err = rand.Read(nonce); err != nil {
		return nil, fmt.Errorf("generate nonce: %w", err)
	}

	// GCM appends the tag to the ciphertext
	sealed := gcm.Seal(nil, nonce, []byte(plaintext), nil)
	ctLen := len(sealed) - gcm.Overhead()
	ciphertext := sealed[:ctLen]
	tag := sealed[ctLen:]

	return &EncryptedPayload{
		Alg:   "AES-256-GCM",
		Data:  base64.StdEncoding.EncodeToString(ciphertext),
		Nonce: base64.StdEncoding.EncodeToString(nonce),
		Tag:   base64.StdEncoding.EncodeToString(tag),
	}, nil
}

// AesGcmDecrypt decrypts an AES-256-GCM encrypted payload.
func AesGcmDecrypt(keyBase64URL string, payload *EncryptedPayload) (string, error) {
	key, err := FromBase64URL(keyBase64URL)
	if err != nil {
		return "", fmt.Errorf("decode key: %w", err)
	}
	data, err := base64.StdEncoding.DecodeString(payload.Data)
	if err != nil {
		return "", fmt.Errorf("decode data: %w", err)
	}
	nonce, err := base64.StdEncoding.DecodeString(payload.Nonce)
	if err != nil {
		return "", fmt.Errorf("decode nonce: %w", err)
	}
	tag, err := base64.StdEncoding.DecodeString(payload.Tag)
	if err != nil {
		return "", fmt.Errorf("decode tag: %w", err)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("new cipher: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("new gcm: %w", err)
	}

	// GCM expects ciphertext+tag concatenated
	sealed := append(data, tag...)
	plaintext, err := gcm.Open(nil, nonce, sealed, nil)
	if err != nil {
		return "", fmt.Errorf("decrypt: %w", err)
	}

	return string(plaintext), nil
}

// ── Ed25519 Signing & Verification ───────────────────────────

// GenerateEd25519KeyPair generates an Ed25519 signing keypair.
func GenerateEd25519KeyPair() (publicKeyBase64URL, privateKeyBase64URL string, err error) {
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return "", "", fmt.Errorf("generate ed25519 key: %w", err)
	}
	return ToBase64URL(pub), ToBase64URL(priv), nil
}

// Ed25519Sign signs a message with Ed25519. Returns base64url signature.
func Ed25519Sign(privateKeyBase64URL string, message []byte) (string, error) {
	privBytes, err := FromBase64URL(privateKeyBase64URL)
	if err != nil {
		return "", fmt.Errorf("decode private key: %w", err)
	}
	sig := ed25519.Sign(ed25519.PrivateKey(privBytes), message)
	return ToBase64URL(sig), nil
}

// Ed25519Verify verifies an Ed25519 signature. Returns true if valid.
func Ed25519Verify(publicKeyBase64URL string, message []byte, signatureBase64URL string) (bool, error) {
	pubBytes, err := FromBase64URL(publicKeyBase64URL)
	if err != nil {
		return false, fmt.Errorf("decode public key: %w", err)
	}
	sig, err := FromBase64URL(signatureBase64URL)
	if err != nil {
		return false, fmt.Errorf("decode signature: %w", err)
	}
	if len(pubBytes) != ed25519.PublicKeySize {
		return false, errors.New("invalid public key size")
	}
	return ed25519.Verify(ed25519.PublicKey(pubBytes), message, sig), nil
}

File: src/rust/src/crypto.rs
//! HARP cryptographic primitives: AES-256-GCM, X25519, HKDF, Ed25519, Base64url, SHA-256.
//! Adapted from the HARP reference implementation (harp-samples).

use aes_gcm::{aead::Aead, Aes256Gcm, KeyInit, Nonce};
use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
use ed25519_dalek::{Signer, SigningKey, Verifier, VerifyingKey};
use hkdf::Hkdf;
use rand::rngs::OsRng;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use x25519_dalek::{PublicKey as X25519PubKey, StaticSecret};

const AES_KEY_SIZE: usize = 32;
const AES_NONCE_SIZE: usize = 12;

/// Encrypted payload (AES-256-GCM) with base64-encoded fields.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EncryptedPayload {
    pub alg: String,
    pub data: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nonce: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tag: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub aad: Option<String>,
}

// ── Base64url ────────────────────────────────────────────────

/// Encode bytes to base64url (no padding).
pub fn to_base64url(bytes: &[u8]) -> String {
    URL_SAFE_NO_PAD.encode(bytes)
}

/// Decode base64url (with or without padding).
pub fn from_base64url(s: &str) -> Result<Vec<u8>, base64::DecodeError> {
    URL_SAFE_NO_PAD.decode(s.trim_end_matches('='))
}

// ── SHA-256 ──────────────────────────────────────────────────

/// Compute lowercase hex SHA-256 of a UTF-8 string.
pub fn sha256_hex(s: &str) -> String {
    let hash = Sha256::digest(s.as_bytes());
    hex::encode(hash)
}

// ── HKDF-SHA256 ──────────────────────────────────────────────

/// Derive a key using HKDF-SHA256 (RFC 5869).
pub fn hkdf_derive(ikm: &[u8], salt: &[u8], info: &[u8], length: usize) -> Result<Vec<u8>, String> {
    let hkdf = Hkdf::<Sha256>::new(if salt.is_empty() { None } else { Some(salt) }, ikm);
    let mut okm = vec![0u8; length];
    hkdf.expand(info, &mut okm).map_err(|e| format!("hkdf expand: {e}"))?;
    Ok(okm)
}

// ── X25519 ECDH Key Exchange ─────────────────────────────────

/// X25519 keypair for ECDH key agreement.
pub struct X25519KeyPair {
    pub public_key: String,  // base64url (raw 32 bytes)
    pub private_key: String, // base64url (raw 32 bytes)
}

/// Generate an X25519 keypair for ECDH key agreement.
pub fn generate_x25519_keypair() -> X25519KeyPair {
    let secret = StaticSecret::random_from_rng(OsRng);
    let public = X25519PubKey::from(&secret);
    X25519KeyPair {
        public_key: to_base64url(public.as_bytes()),
        private_key: to_base64url(secret.as_bytes()),
    }
}

/// Derive a shared AES-256 key from X25519 ECDH + HKDF-SHA256.
pub fn derive_shared_key(my_private_base64url: &str, peer_public_base64url: &str) -> Result<String, String> {
    let priv_bytes = from_base64url(my_private_base64url).map_err(|e| format!("decode privkey: {e}"))?;
    let pub_bytes = from_base64url(peer_public_base64url).map_err(|e| format!("decode pubkey: {e}"))?;

    let mut priv_arr = [0u8; 32];
    let mut pub_arr = [0u8; 32];
    if priv_bytes.len() != 32 || pub_bytes.len() != 32 {
        return Err("keys must be 32 bytes".into());
    }
    priv_arr.copy_from_slice(&priv_bytes);
    pub_arr.copy_from_slice(&pub_bytes);

    let secret = StaticSecret::from(priv_arr);
    let public = X25519PubKey::from(pub_arr);
    let shared_secret = secret.diffie_hellman(&public);

    let info = b"HARP-E2E-AES256GCM";
    let derived = hkdf_derive(shared_secret.as_bytes(), &[], info, AES_KEY_SIZE)?;
    Ok(to_base64url(&derived))
}

// ── AES-256-GCM Encrypt/Decrypt ──────────────────────────────

/// Encrypt plaintext with AES-256-GCM.
pub fn aes_gcm_encrypt(key_base64url: &str, plaintext: &str) -> Result<EncryptedPayload, String> {
    let key = from_base64url(key_base64url).map_err(|e| format!("decode key: {e}"))?;
    if key.len() != AES_KEY_SIZE {
        return Err(format!("key must be {AES_KEY_SIZE} bytes, got {}", key.len()));
    }

    let cipher = Aes256Gcm::new_from_slice(&key).map_err(|e| format!("new cipher: {e}"))?;

    let mut nonce_bytes = [0u8; AES_NONCE_SIZE];
    use rand::RngCore;
    OsRng.fill_bytes(&mut nonce_bytes);
    let nonce = Nonce::from_slice(&nonce_bytes);

    let ciphertext = cipher.encrypt(nonce, plaintext.as_bytes()).map_err(|e| format!("encrypt: {e}"))?;

    // AES-GCM appends the 16-byte tag
    let ct_len = ciphertext.len() - 16;
    let ct = &ciphertext[..ct_len];
    let tag = &ciphertext[ct_len..];

    Ok(EncryptedPayload {
        alg: "AES-256-GCM".into(),
        data: base64::engine::general_purpose::STANDARD.encode(ct),
        nonce: Some(base64::engine::general_purpose::STANDARD.encode(nonce_bytes)),
        tag: Some(base64::engine::general_purpose::STANDARD.encode(tag)),
        aad: None,
    })
}

/// Decrypt an AES-256-GCM encrypted payload.
pub fn aes_gcm_decrypt(key_base64url: &str, payload: &EncryptedPayload) -> Result<String, String> {
    let key = from_base64url(key_base64url).map_err(|e| format!("decode key: {e}"))?;
    let data = base64::engine::general_purpose::STANDARD.decode(&payload.data)
        .map_err(|e| format!("decode data: {e}"))?;
    let nonce_bytes = base64::engine::general_purpose::STANDARD.decode(payload.nonce.as_deref().unwrap_or(""))
        .map_err(|e| format!("decode nonce: {e}"))?;
    let tag = base64::engine::general_purpose::STANDARD.decode(payload.tag.as_deref().unwrap_or(""))
        .map_err(|e| format!("decode tag: {e}"))?;

    let cipher = Aes256Gcm::new_from_slice(&key).map_err(|e| format!("new cipher: {e}"))?;
    let nonce = Nonce::from_slice(&nonce_bytes);

    let mut sealed = data;
    sealed.extend_from_slice(&tag);
    let plaintext = cipher.decrypt(nonce, sealed.as_ref()).map_err(|e| format!("decrypt: {e}"))?;

    String::from_utf8(plaintext).map_err(|e| format!("utf8: {e}"))
}

// ── Ed25519 Signing & Verification ───────────────────────────

/// Ed25519 keypair.
pub struct Ed25519KeyPair {
    pub public_key: String,  // base64url (raw 32 bytes)
    pub private_key: String, // base64url (raw 32 bytes)
}

/// Generate an Ed25519 signing keypair.
pub fn generate_ed25519_keypair() -> Ed25519KeyPair {
    let signing_key = SigningKey::generate(&mut OsRng);
    let verifying_key = signing_key.verifying_key();
    Ed25519KeyPair {
        public_key: to_base64url(verifying_key.as_bytes()),
        private_key: to_base64url(signing_key.as_bytes()),
    }
}

/// Sign a message with Ed25519. Returns base64url signature.
pub fn ed25519_sign(private_key_base64url: &str, message: &[u8]) -> Result<String, String> {
    let priv_bytes = from_base64url(private_key_base64url).map_err(|e| format!("decode key: {e}"))?;
    let mut key_arr = [0u8; 32];
    if priv_bytes.len() != 32 {
        return Err("private key must be 32 bytes".into());
    }
    key_arr.copy_from_slice(&priv_bytes);
    let signing_key = SigningKey::from_bytes(&key_arr);
    let sig = signing_key.sign(message);
    Ok(to_base64url(&sig.to_bytes()))
}

/// Verify an Ed25519 signature. Returns true if valid.
pub fn ed25519_verify(public_key_base64url: &str, message: &[u8], signature_base64url: &str) -> Result<bool, String> {
    let pub_bytes = from_base64url(public_key_base64url).map_err(|e| format!("decode key: {e}"))?;
    let sig_bytes = from_base64url(signature_base64url).map_err(|e| format!("decode sig: {e}"))?;

    let mut key_arr = [0u8; 32];
    if pub_bytes.len() != 32 { return Err("public key must be 32 bytes".into()); }
    key_arr.copy_from_slice(&pub_bytes);

    let mut sig_arr = [0u8; 64];
    if sig_bytes.len() != 64 { return Err("signature must be 64 bytes".into()); }
    sig_arr.copy_from_slice(&sig_bytes);

    let verifying_key = VerifyingKey::from_bytes(&key_arr).map_err(|e| format!("import key: {e}"))?;
    let signature = ed25519_dalek::Signature::from_bytes(&sig_arr);
    Ok(verifying_key.verify(message, &signature).is_ok())
}
